firebase_admin/app_check.py

```diff
@@ -0,0 +1,132 @@
+
+    Args:
+        token: A token from App Check.
+        app: An App instance (optional).
+
+    Returns:
+        Dict[str, Any]: A token's decoded claims
+        if the App Check token is valid; otherwise, a rejected promise.
+    """
+    return _get_app_check_service(app).verify_token(token)
+
+class _AppCheckService:
+    """Service class that implements Firebase App Check functionality."""
+
+    _APP_CHECK_ISSUER = 'https://firebaseappcheck.googleapis.com/'
+    _JWKS_URL = 'https://firebaseappcheck.googleapis.com/v1/jwks'
+    _project_id = None
+
+    def __init__(self, app):
+        # Validate and store the project_id to validate the JWT claims
+        self._project_id = app.project_id
+        if not self._project_id:
+            raise ValueError(
+                'Project ID is required to access App Check service. Either set the '
+                'projectId option, or use service account credentials. Alternatively, set the '
+                'GOOGLE_CLOUD_PROJECT environment variable.')
+
+    def verify_token(self, token: str) -> Dict[str, Any]:
+        """Verifies a Firebase App Check token."""
+        _Validators.check_string("app check token", token)
+
+        # Obtain the Firebase App Check Public Keys
+        # Note: It is not recommended to hard code these keys as they rotate,
+        # but you should cache them for up to 6 hours.
+        # Default lifespan is 300 seconds (5 minutes) so we change it to 21600 seconds (6 hours).
+        jwks_client = PyJWKClient(self._JWKS_URL, lifespan=21600)
+        signing_key = jwks_client.get_signing_key_from_jwt(token)
+        self._has_valid_token_headers(jwt.get_unverified_header(token))
+        verified_claims = self._decode_and_verify(token, signing_key.get('key'))
+
+        # The token's subject will be the app ID, you may optionally filter against
+        # an allow list
+        verified_claims['app_id'] = verified_claims.get('sub')
+        return verified_claims
+
+    def _has_valid_token_headers(self, headers: Any) -> None:
+        """Checks whether the token has valid headers for App Check."""
+        # Ensure the token's header has type JWT
+        if headers.get('typ') != 'JWT':
+            raise ValueError("The provided App Check token has an incorrect type header")
+        # Ensure the token's header uses the algorithm RS256
+        algorithm = headers.get('alg')
+        if algorithm != 'RS256':
+            raise ValueError(
+                'The provided App Check token has an incorrect algorithm. '
+                f'Expected RS256 but got {algorithm}.'
+                )
+
+    def _decode_token(self, token: str, signing_key: str, algorithms: List[str]) -> Dict[str, Any]:
+        """Decodes the JWT received from App Check."""
+        payload = {}
+        try:
+            payload = jwt.decode(
+                token,
+                signing_key,
+                algorithms
+            )
+        except (DecodeError, InvalidKeyError):
```

Comment:
having difficulty testing these errors

Reply:
Yeah.. I think to test these we will have to mock `jwt.decode()` (or wrap it in a helper class and throw from there). Let's continue with the rest and come back to this later.